axelrod/strategies/finite_state_machines.py

```diff
@@ -1,7 +1,238 @@
+            visited[destination] = False
+
+    # This is what we'll recurse on.  visited dict is shared between calls.
+    def recurse(at_node):
+        visited[at_node] = True
+        record = 1  # Count the nodes, not the edges.
+        for successor in edges[at_node]:
+            if visited[successor]:
+                return float("inf")
+            successor_length = recurse(successor)
+            if successor_length == float("inf"):
+                return float("inf")
+            if record < successor_length + 1:
+                record = successor_length + 1
+        return record
+
+    return recurse(starting_at)
+
+
+def get_memory_from_transitions(transitions: TransitionDict,
+                                initial_state: int = None) -> int:
+    """This function calculates the memory of an FSM from the transitions.
+
+    Assume that transitions are a dict with entries like
+    (state, last_opponent_action): (next_state, next_action)
+
+    We first break down the transitions into memits (see above).  We also create
+    a graph of memits, where the successor to a given memit are all possible
+    memits that could occur in the memory immediately before the given memit.
+
+    Then we pair up memits with different states, but same in and out actions.
+    These represent points in time that we can't determine which state we're in.
+    We also create a graph of memit-pairs, where memit-pair, Y, succedes a
+    memit-pair, X, if the two memits in X are succeded by the two memits in Y.
+    These edges reperesent consecutive points in time that we can't determine
+    which state we're in.
+
+    Then for all memit-pairs that disagree, in the sense that they imply
+    different next_action, we find the longest chain starting at that
+    memit-pair.  [If a loop is encountered then this will be infinite.]  We take
+    the maximum over all such memit-pairs.  This represents the longest possible
+    chain of memory for which we wouldn't know what to do next.  We return this.
+    """
+    # If initial_state is set, use this to determine which transitions are
+    # reachable from the initial_state and restrict to those.
+    if initial_state is not None:
+        transitions = get_accessible_transitions(transitions, initial_state)
+
+    # Get the incoming actions for each state.
+    incoming_action_by_state = defaultdict(set)  # type: DefaultDict[int, Set[Action]]
+    for trans in transition_iterator(transitions):
+        incoming_action_by_state[trans.next_state].add(trans.next_action)
+
+    # Keys are starting memit, and values are all possible terminal memit.
+    # Will walk backwards through the graph.
+    memit_edges = defaultdict(set)  # type: DefaultDict[Memit, Set[Memit]]
+    for trans in transition_iterator(transitions):
+        # Since all actions are out-paths for each state, add all of these.
+        # That is to say that your opponent could do anything
+        for out_action in ALL_ACTIONS:
```

Comment:
Since this is then only place where `ALL_ACTIONS` is used I think it's better as a default kwarg to the function rather than a global.